modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_navigation/js/AssetCategoriesNavigationTreeView.js

```diff
@@ -15,13 +15,44 @@
 import {Treeview} from 'frontend-js-components-web';
 import React from 'react';
 
-const AssetCategoriesNavigationTreeView = ({categories, namespace}) => {
+function buildNodes(vocabularies, categories) {
+	const nodes = [];
+
+	vocabularies.forEach((vocabulary) => {
+		const child = {
+			...vocabulary,
+			children: [],
+			id: vocabulary.vocabularyId,
+		};
+
+		categories.forEach((category) => {
+			if (category.vocabularyId === child.id) {
+				child.children.push({
+					...category,
+					id: category.categoryId,
+				});
+			}
+		});
+
+		nodes.push(child);
+	});
+
+	return nodes;
```

Comment:
@julien I don't know if your next iteration is going to have anything like this in it, but in general a pattern like:

```js
const nodes = [];

vocabularies.forEach((_) => {
    // ...

    nodes.push(child);
});

return nodes;
```

Usually just wants to be:

```js
return vocabularies.map((_) => {
    // ...

    return child;
});
```

You know, functional/immutable generally preferable to imperative/mutable and all that. Depending on how far you want to go down that path, in this one, you have a nested `forEach` that has the same kind of pattern; could be:

```js
const child = {
	...vocabulary,
	children: categories
		.map((category) => {
			if (category.vocabularyId === child.id) {
				return {
					...category,
					id: category.categoryId,
				};
			}
		})
		.filter(Boolean),
};
```

(ie. using `filter(Boolean)` to skip empty `map` results)

All together, that would be:

```js
function buildNodes(vocabularies, categories) {
	return vocabularies.map((vocabulary) => ({
		...vocabulary,
		children: categories
			.map((category) => {
				if (category.vocabularyId === child.id) {
					return {
						...category,
						id: category.categoryId,
					};
				}
			})
			.filter(Boolean),
	}));
}
```

Having said that, are the multiple levels of nesting possible? Handling those would imply the need for a recursive solution (or equivalent).

Reply:
Copy pasted this over to [the new PR](https://github.com/mateomustapic/liferay-portal/pull/10), so may as well ignore the above.

Reply:
@wincent, that's great, thanks for the help. 
Yes, we will need something recursive because categories can also have nested children than can also have nested children, but I think that based on your feedback we can come up with something that works.